### ISSUE TITLE
OpenSSL on Windows

### ISSUE THREAD

**Issue by aleks-f** [pocoproject/poco#686]
@obiltschnig:

Three questions (diff summary of the proposal at the bottom):

1. Since static libraries do not link at build time are [these entries](https://github.com/pocoproject/poco/blob/develop/Crypto/Crypto.progen#L18-L21) in progen file needed?

2. Since the ShiningLight Win OpenSSL provides debug/release and static/dynamic runtime VC libraries to link with, why do we link with the ones in /lib directory? Shouldn't it be changed to link with lib/VC binaries? 

3. Can we add default windows OpenSSL installation directories (C:\\OpenSSL-Win32 and C:\\OpenSSL-Win64) to the list of include and additional library directories? While this would still result in compile/link errors for users who have OpenSSL installed elsewhere (we did not do any harm to them because it would fail anyway), it will at least work for those who install in default directories.

``` diff
$ git diff Crypto/Crypto.progen
diff --git a/Crypto/Crypto.progen b/Crypto/Crypto.progen
index a7f3cfa..1fb7866 100644
--- a/Crypto/Crypto.progen
+++ b/Crypto/Crypto.progen
@@ -8,16 +8,16 @@ vc.project.platforms = Win32, x64, WinCE
 vc.project.configurations = debug_shared, release_shared, debug_static_mt, rele
 vc.project.prototype = ${vc.project.name}_vs90.vcproj
 vc.project.compiler.include = ..\\Foundation\\include
+vc.project.compiler.include.Win32 = C:\\OpenSSL-Win32\\include
+vc.project.compiler.include.x64 = C:\\OpenSSL-Win64\\include
 vc.project.compiler.defines =
 vc.project.compiler.defines.shared = ${vc.project.name}_EXPORTS
 vc.project.compiler.defines.debug_shared = ${vc.project.compiler.defines.shared
 vc.project.compiler.defines.release_shared = ${vc.project.compiler.defines.shar
 vc.project.linker.dependencies = ws2_32.lib iphlpapi.lib
-vc.project.linker.dependencies.debug_shared = libeay32.lib ssleay32.lib
-vc.project.linker.dependencies.release_shared = libeay32.lib ssleay32.lib
-vc.project.linker.dependencies.debug_static_md = libeay32mdd.lib ssleay32mdd.li
-vc.project.linker.dependencies.release_static_md = libeay32md.lib ssleay32md.li
-vc.project.linker.dependencies.debug_static_mt = libeay32mtd.lib ssleay32mtd.li
-vc.project.linker.dependencies.release_static_mt = libeay32mt.lib ssleay32mt.li
+vc.project.linker.dependencies.debug_shared = libeay32MDd.lib ssleay32MDd.lib
+vc.project.linker.dependencies.release_shared = libeay32MD.lib ssleay32MD.lib
+vc.project.linker.additionalOptions.Win32 = /LIBPATH:"C:\\OpenSSL-Win32\\lib\\VC"
+vc.project.linker.additionalOptions.x64 = /LIBPATH:"C:\\OpenSSL-Win64\\lib\\VC"
 vc.solution.create = true
 vc.solution.include = testsuite\\TestSuite
(END)
```

**Comment by aleks-f** [comment]
An additional thought: OpenSSL binaries are built with VS2008, so when Poco is built with another VS version, POCO and OpenSSL will have different heaps, which likely causes this [Dr. Memory](http://drmemory.org/docs/page_invarg.html) error:
```
~~Dr.M~~ Error #29: INVALID HEAP ARGUMENT to Windows API routine: invalid heap HANDLE 0x00000000
~~Dr.M~~ # 0 replace_RtlQueryHeapInformation                                 [d:\drmemory_package\common\alloc_replace.c:3652]
~~Dr.M~~ # 1 ntdll.dll!RtlQueryProcessHeapInformation                       +0x287    (0x779241fd <ntdll.dll+0xa41fd>)
~~Dr.M~~ # 2 ntdll.dll!RtlUlonglongByteSwap                                 +0x5c3b   (0x779030dc <ntdll.dll+0x830dc>)
~~Dr.M~~ # 3 KERNEL32.dll!Heap32First                                       +0x55     (0x761e5cf9 <KERNEL32.dll+0x95cf9>)
~~Dr.M~~ # 4 LIBEAY32.dll!RAND_poll                                         +0x531    (0x0634dd62 <LIBEAY32.dll+0x5dd62>)
~~Dr.M~~ # 5 LIBEAY32.dll!RAND_SSLeay                                       +0x4d3    (0x0634ca14 <LIBEAY32.dll+0x5ca14>)
~~Dr.M~~ # 6 LIBEAY32.dll!RAND_SSLeay                                       +0x8b2    (0x0634cdf3 <LIBEAY32.dll+0x5cdf3>)
~~Dr.M~~ # 7 SSLEAY32.dll!SSL_CTX_new                                       +0x2a4    (0x03f95fb5 <SSLEAY32.dll+0x25fb5>)
~~Dr.M~~ # 8 PocoNetSSLd.dll!Poco::Net::Context::createSSLContext            [c:\poco-1.5.3\netssl_openssl\src\context.cpp:318]
~~Dr.M~~ # 9 PocoNetSSLd.dll!Poco::Net::Context::Context                     [c:\poco-1.5.3\netssl_openssl\src\context.cpp:51]
~~Dr.M~~ #10 PocoNetSSLd.dll!Poco::Net::SSLManager::initDefaultContext       [c:\poco-1.5.3\netssl_openssl\src\sslmanager.cpp:255]
~~Dr.M~~ #11 PocoNetSSLd.dll!Poco::Net::SSLManager::defaultServerContext     [c:\poco-1.5.3\netssl_openssl\src\sslmanager.cpp:117]
~~Dr.M~~ #12 PocoNetSSLd.dll!Poco::Net::SecureServerSocket::SecureServerSocket [c:\poco-1.5.3\netssl_openssl\src\secureserversocket.cpp:69]
```

Likely a false positive because, if we do not allocate on one heap and free on the other, we should be good, but it's a thin line ...

**Comment by aleks-f** [comment]
Resolved with [#780](https://github.com/pocoproject/poco/issues/780) for [1.7.0](https://github.com/pocoproject/poco/milestones/Release%201.7.0)
